### PR TITLE
 Lint against single-use lifetime names

### PR DESCRIPTION
**Pull request by gaurikholkar-zz**
This is a fix for #44752 

TO-DO

- [x] change lint message
- [x] add ui tests 

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/lint/builtin.rs [hunk 1]
```diff
@@ -216,48 +216,53 @@ declare_lint! {
...
+
 /// Does nothing as a lint pass, but registers some `Lint`s
 /// which are used by other parts of the compiler.
 #[derive(Copy, Clone)]
 pub struct HardwiredLints;
 
 impl LintPass for HardwiredLints {
     fn get_lints(&self) -> LintArray {
-        lint_array!(
-            UNUSED_IMPORTS,
```
Comment (oli-obk) [inline src/librustc/lint/builtin.rs]: Unrelated reformatting.

File: src/librustc/middle/resolve_lifetime.rs [hunk 2]
```diff
@@ -32,23 +32,41 @@ use errors::DiagnosticBuilder;
...
+}
+
 #[derive(Clone, Copy, PartialEq, Eq, Hash, RustcEncodable, RustcDecodable, Debug)]
 pub enum Region {
     Static,
-    EarlyBound(/* index */ u32, /* lifetime decl */ DefId),
-    LateBound(ty::DebruijnIndex, /* lifetime decl */ DefId),
-    LateBoundAnon(ty::DebruijnIndex, /* anon index */ u32),
-    Free(DefId, /* lifetime decl */ DefId),
+    EarlyBound(/* index */
```
Comment (oli-obk) [inline src/librustc/middle/resolve_lifetime.rs]: Please move the formatting changes into their own commit or undo them. It's very hard to see the non-formatting changes among them

**Comment by nikomatsakis** [comment]
Regarding the formatting changes, what I usually find helpful is to have the first commit just rustfmt's the whole file, and then start from there. Then when you rebase, if you get conflicts in this commit, you can just take the "master" version of the code indiscriminately and then re-run rustfmt afterwards. (This...mostly works. Sometimes I find code gets duplicated.)

**Comment by gaurikholkar** [comment]
I guess most of it is undone. Will take care of the rest in a separate commit

File: src/librustc/lint/builtin.rs (hunk 1 above)
Comment (gaurikholkar) [reply src/librustc/lint/builtin.rs]: fixed formatting changes @oli-obk

**Comment by gaurikholkar** [comment]
@nikomatsakis @oli-obk I get this error
```
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.24.0-dev running on x86_64-apple-darwin

thread 'rustc' panicked at 'assertion failed: def_id.is_local()', src/librustc/hir/map/mod.rs:309:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `alloc`.

Caused by:
  process didn't exit successfully: `/Users/gkholkar/Desktop/or/rust/build/bootstrap/debug/rustc --crate-name alloc src/liballoc/lib.rs --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=22e63b537c61a4f8 -C extra-filename=-22e63b537c61a4f8 --out-dir /Users/gkholkar/Desktop/or/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/Users/gkholkar/Desktop/or/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps -L dependency=/Users/gkholkar/Desktop/or/rust/build/x86_64-apple-darwin/stage1-std/release/deps --extern core=/Users/gkholkar/Desktop/or/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libcore-a1b407ed3953731d.rlib --extern std_unicode=/Users/gkholkar/Desktop/or/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps/libstd_unicode-aa82a4fba39f1481.rlib` (exit code: 101)
thread 'main' panicked at 'command did not execute successfully: "/Users/gkholkar/Desktop/or/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "-j" "4" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/Users/gkholkar/Desktop/or/rust/src/libstd/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101', src/bootstrap/compile.rs:882:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failed to run: /Users/gkholkar/Desktop/or/rust/build/bootstrap/debug/bootstrap build --stage 1 src/libsyntax
Build completed unsuccessfully in 0:18:18
```

**Comment by gaurikholkar** [comment]
Will fix the ui tests

**Comment by oli-obk** [comment]
They are failing with an ICE: `unelided lifetime in signature`. Can you get a backtrace for such a failure?

**Comment by gaurikholkar** [comment]
Afk for a while. Will check as asap

**Review by gaurikholkar-zz — COMMENTED** [review]
How do I write both the errors occuring on the same line on the .rs files for this test?

**Comment by oli-obk** [comment]
You can write `//~^ ERROR foo` in the line below. If you have even more messages per line, start adding `//~| ERROR bar` below a line with the `^`

**Comment by nikomatsakis** [comment]
So I feel like the new code is generating unexpected errors. I'm going to poke around a bit.

**Comment by nikomatsakis** [comment]
I'm going to open a PR with just the resolve-lifetimes query work, so we can isolate what is effecting what.

**Comment by nikomatsakis** [comment]
Blocked on https://github.com/rust-lang/rust/pull/46657

**Comment by nikomatsakis** [comment]
OK #46657 landed -- @gaurikholkar, can you rebase atop that?

**Comment by gaurikholkar** [comment]
@nikomatsakis at it

**Comment by gaurikholkar** [comment]
@nikomatsakis build passed :)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
This is looking pretty good. I think we need a few more tests. It's probably not quite working right, but I'd be happy to get something landed and try to refine later.

**Comment by nikomatsakis** [comment]
Oh, in my review, I realize I wasn't clear. The "Nit: formatting" bits was just meant to say that these changes weren't *necessary*, though also not harmful. Ideally they would be isolated into a separate commit (or else pulled out).

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**